src/gui/qgsmapcanvas.cpp

```diff
@@ -778,8 +796,26 @@ QgsRectangle QgsMapCanvas::imageRect( const QImage& img, const QgsMapSettings& m
 
 void QgsMapCanvas::mapUpdateTimeout()
 {
-  const QImage& img = mJob->renderedImage();
-  mMap->setContent( img, imageRect( img, mJob->mapSettings() ) );
+  QImage img = mJob->renderedImage();
+  magnifyImage( img );
+
+  mMap->setContent( img, imageRect( img, mSettings ) );
+}
+
+void QgsMapCanvas::magnifyImage( QImage &img )
+{
+  if ( mMagnificationLevel != 1.0 )
+  {
+    img = img.scaled( img.width() * mMagnificationLevel,
```

Comment:
I took a look into QgsMapSettings to see what I can do about the non pixelated zoom, and I have some questions.  

I noticed that the mScaleFactor of [QgsRenderContext](https://github.com/qgis/QGIS/blob/master/src/core/qgsrendercontext.cpp) depends on the DPI of QgsMapSettings. This factor is used to scale line widths, point marker sizes, ... So, when I increase the DPI through QgsMapSettings, the extent of map canvas and the size in pixels of the rendered image are the same, but labels, lines and markers are bigger. This also implies that for the same extent with different DPI, labels are at different position. Is this normal? 

Similarly, if I increase the outputSize through QgsMapSettings, the rendered image is bigger (in terms of pixels). However, the size of markers (for example) doesn't change. So, more I increase the outputSize, more markers, labels, .. seem little within the rendered image. Is this normal?

Let me know if I missed something.